### PR TITLE
feat(estimate_size): `EstimateSize` for scalars

### LINKED ISSUE #9371: implement proc macro for `EstimateSize`

### PR DESCRIPTION
**Pull request by yuhao-su**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- add proc macro to derive `EstimateSize`
- add `EstimateSize` for scalars.

close #9371

## Checklist For Contributors

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/common/proc_macro/src/config.rs [hunk 1]
```diff
@@ -12,6 +12,26 @@
...
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
```
Comment (liurenjie1024) [inline src/common/proc_macro/src/config.rs:33]: Why adding this?

File: src/common/proc_macro/src/lib.rs [hunk 2]
```diff
@@ -48,10 +54,172 @@ mod config;
...
+pub fn override_config(input: TokenStream) -> TokenStream {
     let input = parse_macro_input!(input);
 
     let gen = config::produce_override_config(input);
 
     gen.into()
 }
+
+#[proc_macro_derive(EstimateSize)]
+pub fn derive_estimate_size(input: TokenStream) -> TokenStream {
```
Comment (liurenjie1024) [inline src/common/proc_macro/src/lib.rs:66]: I think `GetSize` crate has implemented similar function? If so, do we really need to maintain one for this?

File: src/common/src/types/num256.rs [hunk 3]
```diff
@@ -360,6 +361,12 @@ impl<'a> From<Int256Ref<'a>> for arrow_buffer::i256 {
     }
 }
 
+impl EstimateSize for Int256 {
+    fn estimated_heap_size(&self) -> usize {
+        mem::size_of::<i128>() * 2
```
Comment (liurenjie1024) [inline src/common/src/types/num256.rs:366]: This is quite confusing to me, I think this should return 0.

File: src/common/proc_macro/src/config.rs (hunk 1 above)
Comment (yuhao-su) [reply src/common/proc_macro/src/config.rs:33]: It's from GetSize

Comment (yuhao-su) [reply src/common/proc_macro/src/config.rs:33]: Oh may bad..

File: src/common/proc_macro/src/estimate_size.rs [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+                }
+            }
+            if found {
+                continue;
+            }
+            type_param.bounds.push(syn::parse_quote!(EstimateSize));
+        }
+    }
+    generics
+}
```
Comment (liurenjie1024) [inline src/common/proc_macro/src/estimate_size.rs:150]: Can we also port tests?

**Review by liurenjie1024 — COMMENTED** [review]
Generally LGTM

File: src/common/proc_macro/src/estimate_size.rs (hunk 4 above)
Comment (yuhao-su) [reply src/common/proc_macro/src/estimate_size.rs:150]: It simply won't compile if there is a bug in proc macro

**Review by liurenjie1024 — APPROVED** [review]

**Pull request merged**